velox/experimental/gpu/CudaMemMeter.cu

```diff
@@ -0,0 +1,256 @@
+                                   blockSize,
+                                   bytes,
+                                   streamBytes,
+                                   nRepeats,
+                                   nStreams,
+                                   oneWay,
+                                   multiplier]() {
+      auto thisDev = dev;
+      checkCuda(cudaSetDevice(thisDev));
+      // allocate pinned host memory and device memory
+      float *a, *d_a;
+      printf("%dKB\n", bytes / 1024);
+      checkCuda(cudaMallocHost((void**)&a, bytes)); // host pinned
+      checkCuda(cudaMalloc((void**)&d_a, bytes)); // device
+
+      float ms; // elapsed time in milliseconds
+      float singleMs = 0;
+      float async1Ms = 0;
+      float async2Ms = 0;
+
+      // create events and streams
+      cudaEvent_t startEvent, stopEvent, dummyEvent;
+      cudaStream_t* stream =
+          (cudaStream_t*)malloc(nStreams * sizeof(cudaStream_t));
+      checkCuda(cudaEventCreate(&startEvent));
+      checkCuda(cudaEventCreate(&stopEvent));
+      checkCuda(cudaEventCreate(&dummyEvent));
+      for (int i = 0; i < nStreams; ++i)
+        checkCuda(cudaStreamCreate(&stream[i]));
+
+      // baseline case - sequential transfer and execute
+      memset(a, 0, bytes);
+      for (int repeat = 0; repeat < nRepeats; ++repeat) {
+        checkCuda(cudaEventRecord(startEvent, 0));
+        checkCuda(cudaMemcpy(d_a, a, bytes, cudaMemcpyHostToDevice));
+        kernel<<<n / blockSize, blockSize>>>(d_a, 0);
+        if (!oneWay) {
+          checkCuda(cudaMemcpy(a, d_a, bytes, cudaMemcpyDeviceToHost));
+        }
+        checkCuda(cudaEventRecord(stopEvent, 0));
+        checkCuda(cudaEventSynchronize(stopEvent));
+        checkCuda(cudaEventElapsedTime(&ms, startEvent, stopEvent));
+        singleMs += ms;
+      }
+      float volume = bytes * nRepeats;
+      printf(
+          "%d at %d: Time for sequential transfer and execute (ms): %f %f GB/s\n",
+          thisDev,
+          millis(start),
+          singleMs,
+          multiplier * (volume / (1 << 30)) / (singleMs / 1000));
+
+      // asynchronous version 1: loop over {copy, kernel, copy}
+      memset(a, 0, bytes);
+      for (int repeat = 0; repeat < nRepeats; ++repeat) {
+        checkCuda(cudaEventRecord(startEvent, 0));
+        for (int i = 0; i < nStreams; ++i) {
+          int offset = i * streamSize;
+          checkCuda(cudaMemcpyAsync(
+              &d_a[offset],
+              &a[offset],
+              streamBytes,
+              cudaMemcpyHostToDevice,
+              stream[i]));
+          kernel<<<streamSize / blockSize, blockSize, 0, stream[i]>>>(
+              d_a, offset);
+          if (!oneWay) {
+            checkCuda(cudaMemcpyAsync(
+                &a[offset],
+                &d_a[offset],
+                streamBytes,
+                cudaMemcpyDeviceToHost,
+                stream[i]));
+          }
+        }
+        checkCuda(cudaEventRecord(stopEvent, 0));
```

Comment:
Here we are only waiting for stream 0, we should wait for all streams, the nvidia example also has this problem.